File: project/app/src/main/java/org/owntracks/android/ui/status/StatusViewModel.java
package org.owntracks.android.ui.status;

import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.os.Build;
import android.os.Bundle;
import android.os.PowerManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.Bindable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import org.greenrobot.eventbus.Subscribe;
import org.owntracks.android.BR;
import org.owntracks.android.data.EndpointState;
import org.owntracks.android.services.MessageProcessor;
import org.owntracks.android.support.Events;
import org.owntracks.android.ui.base.viewmodel.BaseViewModel;
import org.owntracks.android.ui.status.logs.LogViewerActivity;

import java.util.Date;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.android.scopes.ActivityScoped;
import timber.log.Timber;


@ActivityScoped
public class StatusViewModel extends BaseViewModel<StatusMvvm.View> implements StatusMvvm.ViewModel<StatusMvvm.View> {
    private final Context context;
    private EndpointState endpointState;
    private String endpointMessage;

    private Date serviceStarted;
    private long locationUpdated;
    private int queueLength;
    private final MutableLiveData<Boolean> isDozeWhitelisted = new MutableLiveData<>();

    @Inject
    public StatusViewModel(@ApplicationContext Context context) {
        this.context = context;
    }

    public void attachView(@Nullable Bundle savedInstanceState, @NonNull StatusMvvm.View view) {
        super.attachView(savedInstanceState, view);
    }

    @Override
    @Bindable
    public EndpointState getEndpointState() {
        return endpointState;
    }

    @Override
    @Bindable
    public String getEndpointMessage() {
        return endpointMessage;
    }

    @Override
    @Bindable
    public int getEndpointQueue() {
        return queueLength;
    }

    @Override
    @Bindable
    public Date getServiceStarted() {
        return serviceStarted;
    }

    @Override
    public LiveData<Boolean> getDozeWhitelisted() {
        return isDozeWhitelisted;
    }

    @Override
    public void refreshDozeModeWhitelisted() {
        isDozeWhitelisted.postValue(isIgnoringBatteryOptimizations());
    }

    private boolean isIgnoringBatteryOptimizations() {
        return Build.VERSION.SDK_INT < Build.VERSION_CODES.M ||
                ((PowerManager) context.getApplicationContext().getSystemService(Context.POWER_SERVICE)).isIgnoringBatteryOptimizations(context.getApplicationContext().getPackageName());
    }

    @Override
    @Bindable
    public long getLocationUpdated() {
        return locationUpdated;
    }

    @Subscribe(sticky = true)
    public void onEvent(EndpointState e) {
        this.endpointState = e;
        this.endpointMessage = e.getMessage();
        notifyPropertyChanged(BR.endpointState);
        notifyPropertyChanged(BR.endpointMessage);
    }

    @Subscribe(sticky = true)
    public void onEvent(Events.ServiceStarted e) {
        this.serviceStarted = e.getDate();
        notifyPropertyChanged(BR.serviceStarted);
    }

    @Subscribe(sticky = true)
    public void onEvent(Location l) {
        this.locationUpdated = TimeUnit.MILLISECONDS.toSeconds(l.getTime());
        notifyPropertyChanged(BR.locationUpdated);
    }

    @Subscribe(sticky = true)
    public void onEvent(Events.QueueChanged e) {
        Timber.v("queue changed %s", e.getNewLength());
        this.queueLength = e.getNewLength();
        notifyPropertyChanged(BR.endpointQueue);
    }

    public void viewLogs() {
        Intent intent = new Intent(context, LogViewerActivity.class).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
    }
}

File: project/app/src/main/java/org/owntracks/android/ui/status/StatusViewModel.kt
package org.owntracks.android.ui.status

import android.Manifest.permission.ACCESS_BACKGROUND_LOCATION
import android.Manifest.permission.ACCESS_COARSE_LOCATION
import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.app.Application
import android.content.Context
import android.location.Location
import android.os.Build
import android.os.PowerManager
import androidx.core.content.PermissionChecker
import androidx.core.content.PermissionChecker.PERMISSION_DENIED
import androidx.core.content.PermissionChecker.PERMISSION_GRANTED
import androidx.lifecycle.AndroidViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import java.time.Instant
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.owntracks.android.R
import org.owntracks.android.data.EndpointState
import org.owntracks.android.data.repos.EndpointStateRepo
import org.owntracks.android.data.repos.LocationRepo

@HiltViewModel
class StatusViewModel
@Inject
constructor(
    application: Application,
    endpointStateRepo: EndpointStateRepo,
    locationRepo: LocationRepo
) : AndroidViewModel(application) {
  val endpointState: StateFlow<EndpointState> = endpointStateRepo.endpointState
  val endpointQueueLength: StateFlow<Int> = endpointStateRepo.endpointQueueLength
  val serviceStarted: StateFlow<Instant> = endpointStateRepo.serviceStartedDate
  val currentLocation: StateFlow<Location?> = locationRepo.currentPublishedLocation
  private val powerManager =
      (getApplication<Application>().applicationContext.getSystemService(Context.POWER_SERVICE)
          as PowerManager)
  internal val dozeWhitelisted = MutableStateFlow(false)

  private val mutableLocationPermissions =
      MutableStateFlow(R.string.statusLocationPermissionsUnknown)
  val locationPermissions: StateFlow<Int> = mutableLocationPermissions

  fun refreshLocationPermissions() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
          val permissions =
              listOf(ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION, ACCESS_BACKGROUND_LOCATION).map {
                PermissionChecker.checkSelfPermission(getApplication(), it)
              }
          when (permissions) {
            listOf(PERMISSION_DENIED, PERMISSION_DENIED, PERMISSION_DENIED) ->
                R.string.statusLocationPermissionsNone
            listOf(PERMISSION_GRANTED, PERMISSION_DENIED, PERMISSION_DENIED) ->
                R.string.statusLocationPermissionsCoarseForeground
            listOf(PERMISSION_GRANTED, PERMISSION_GRANTED, PERMISSION_DENIED) ->
                R.string.statusLocationPermissionsFineForeground
            listOf(PERMISSION_GRANTED, PERMISSION_DENIED, PERMISSION_GRANTED) ->
                R.string.statusLocationPermissionsCoarseBackground
            listOf(PERMISSION_GRANTED, PERMISSION_GRANTED, PERMISSION_GRANTED) ->
                R.string.statusLocationPermissionsFineBackground
            else -> R.string.statusLocationPermissionsUnknown
          }
        } else {
          val permissions =
              listOf(ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION).map {
                PermissionChecker.checkSelfPermission(getApplication(), it)
              }
          when (permissions) {
            listOf(PERMISSION_DENIED, PERMISSION_DENIED) -> R.string.statusLocationPermissionsNone
            listOf(PERMISSION_GRANTED, PERMISSION_DENIED) ->
                R.string.statusLocationPermissionsCoarseForeground
            listOf(PERMISSION_GRANTED, PERMISSION_GRANTED) ->
                R.string.statusLocationPermissionsFineForeground
            else -> R.string.statusLocationPermissionsUnknown
          }
        }
        .run { mutableLocationPermissions.value = this }
  }

  fun getDozeWhitelisted(): StateFlow<Boolean> = dozeWhitelisted

  fun refreshDozeModeWhitelisted() {
    dozeWhitelisted.value = isIgnoringBatteryOptimizations()
  }

  private fun isIgnoringBatteryOptimizations(): Boolean {
    return powerManager.isIgnoringBatteryOptimizations(
        getApplication<Application>().applicationContext.packageName)
  }
}
